google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionGroup.java

```diff
@@ -101,4 +109,60 @@ public void getPartitions(
 
     observer.onCompleted();
   }
+
+  public ApiFuture<List<QueryPartition>> getPartitions(long desiredPartitionCount) {
```

Comment:
I wonder if you we could do something like this: https://github.com/googleapis/java-firestore/blob/edaa5395be0353fb261d954429c624623bc4e346/google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentReference.java#L383 and expose an Iterable of QueryPartitions? I assume this won't work because of some threading issues, but it would be nice to re-use an existing pattern.

Reply:
If we go this route we lose the ability to chain the result of this method into a series of queries with the following pattern:

```java
    ApiFuture<List<QuerySnapshot>> allQueries = ApiFutures.transformAsync(
        firestore.collectionGroup(randomColl.getId()).getPartitions(3),
        new ApiAsyncFunction<List<QueryPartition>, List<QuerySnapshot>>() {
          @Override
          public ApiFuture<List<QuerySnapshot>> apply(List<QueryPartition> input) {
            logit("ApiAsyncFunction.apply");
            List<ApiFuture<QuerySnapshot>> futures = new ArrayList<>();
            for (QueryPartition part : input) {
              logit("part = %s", part);
              futures.add(part.createQuery().get());
            }
            return ApiFutures.allAsList(futures);
          }
        },
        MoreExecutors.directExecutor()
    );
    List<QuerySnapshot> querySnapshots = allQueries.get();
```

```
2021-01-08T23:52:57.989  [main] com.google.cloud.firestore.CollectionGroup - transformAsync...
2021-01-08T23:52:58.008  [main] com.google.cloud.firestore.CollectionGroup - transformAsync Complete
2021-01-08T23:52:58.008  [main] com.google.cloud.firestore.CollectionGroup - allQueries.get()...
2021-01-08T23:52:58.111  [Gax-4] com.google.cloud.firestore.CollectionGroup - transform->apply()
2021-01-08T23:52:58.116  [Gax-4] com.google.cloud.firestore.CollectionGroup - ApiAsyncFunction.apply
2021-01-08T23:52:58.117  [Gax-4] com.google.cloud.firestore.CollectionGroup - part = com.google.cloud.firestore.QueryPartition@a9b0a06e
2021-01-08T23:52:58.131  [Gax-4] com.google.cloud.firestore.CollectionGroup - part = com.google.cloud.firestore.QueryPartition@19a03734
2021-01-08T23:52:58.309  [main] com.google.cloud.firestore.CollectionGroup - allQueries.get() complete
```

There is also additional state that has to be tracked because the iterator we return would have to potentially run past the end of the iterator returned by `response.iterateAll()` not to mention the iteration state now has to be split across the hasNext and next methods. (It's possible but actually much more code, so I can swap it in if you want me to).